File: src/mcp_glimmung/browser_inspector.py
"""Browser inspection wrapper used by the MCP tool and CLI.

Browser work runs in the leased test slot's `slot-playwright` pod. mcp-glimmung
talks to it over the Playwright WebSocket protocol through the sibling Node
helper. The MCP host does not run Chromium itself; without an active
test-slot lease there is no browser to drive.

Output discipline: the screenshot PNG is never round-tripped through the
agent's context as base64. The Node helper writes it to a pod-local
tempfile whose lifecycle is owned by the Python wrapper. The wrapper
uploads the bytes to glimmung's `POST /v1/inspections` along with the
full structured report; the MCP tool then returns a compact summary that
references the durable artifact URLs. See glimmung#143 for the design
record.
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
import tempfile
import uuid
from pathlib import Path
from typing import Any


SCRIPT_PATH = Path(__file__).with_name("browser_inspector.mjs")


def _truncate(value: str, limit: int) -> str:
    if len(value) <= limit:
        return value
    return value[: max(limit - 3, 0)].rstrip() + "..."


def inspect_url(
    *,
    url: str,
    playwright_ws_endpoint: str,
    viewport: dict[str, int] | None = None,
    wait_ms: int = 2000,
    timeout_ms: int = 30000,
    full_page: bool = True,
    capture_accessibility: bool = False,
    capture_console: bool = True,
    capture_network: bool = True,
    max_elements: int = 80,
    body_text_limit: int = 4000,
    cookies: list[dict[str, Any]] | None = None,
    extra_http_headers: dict[str, str] | None = None,
    local_storage: dict[str, dict[str, str]] | None = None,
) -> dict[str, Any]:
    """Inspect a URL via the slot-playwright WebSocket endpoint and return
    the report dict plus a path to the captured PNG.

    `playwright_ws_endpoint` is the slot's `slot-playwright` Service URL, e.g.
    `ws://slot-playwright.<slot-name>.svc.cluster.local:3000`. Callers from the
    MCP tool surface obtain it from the active test-slot lease.

    Auth-injection parameters (`cookies`, `extra_http_headers`,
    `local_storage`) seed the Playwright `BrowserContext` before `page.goto`
    so the navigation runs authenticated. The slot-playwright pod itself
    holds no credentials — every auth identity has to come from the
    caller (the session pod). Typical tank-operator pattern is documented
    in the MCP tool's docstring.

    Lifecycle ownership: this function creates a pod-local tempfile for
    the screenshot, passes its path to the Node helper, reads the report
    JSON off the helper's stdout, and returns both to the caller. The
    caller (tools.py) owns uploading the bytes to glimmung and unlinking
    the tempfile in its `finally` block.
    """
    if not playwright_ws_endpoint:
        raise ValueError("playwright_ws_endpoint is required")
    width = int((viewport or {}).get("width", 1440))
    height = int((viewport or {}).get("height", 900))

    # Tempfile lives in the mcp-glimmung pod's /tmp. We do not delete on
    # creation; tools.py is responsible for unlinking in its finally so a
    # failed subprocess run still cleans up.
    fd, screenshot_path = tempfile.mkstemp(suffix=".png", prefix="inspection-")
    os.close(fd)

    payload: dict[str, Any] = {
        "url": url,
        "playwrightWsEndpoint": playwright_ws_endpoint,
        "viewport": {"width": width, "height": height},
        "waitMs": wait_ms,
        "timeoutMs": timeout_ms,
        "fullPage": full_page,
        "captureAccessibility": capture_accessibility,
        "captureConsole": capture_console,
        "captureNetwork": capture_network,
        "maxElements": max_elements,
        "bodyTextLimit": body_text_limit,
        "screenshotPath": screenshot_path,
    }
    if cookies is not None:
        if not isinstance(cookies, list) or any(not isinstance(c, dict) for c in cookies):
            raise ValueError("cookies must be a list of dicts")
        payload["cookies"] = cookies
    if extra_http_headers is not None:
        if not isinstance(extra_http_headers, dict) or any(
            not isinstance(k, str) or not isinstance(v, str)
            for k, v in extra_http_headers.items()
        ):
            raise ValueError("extra_http_headers must be a dict of str -> str")
        payload["extraHttpHeaders"] = extra_http_headers
    if local_storage is not None:
        if not isinstance(local_storage, dict) or any(
            not isinstance(origin, str)
            or not isinstance(items, dict)
            or any(not isinstance(k, str) or not isinstance(v, str) for k, v in items.items())
            for origin, items in local_storage.items()
        ):
            raise ValueError(
                "local_storage must be a dict of origin -> dict of str -> str"
            )
        payload["localStorage"] = local_storage
    env = os.environ.copy()
    if "NODE_PATH" not in env:
        node_modules = Path.cwd() / "node_modules"
        if node_modules.exists():
            env["NODE_PATH"] = str(node_modules)
    proc = subprocess.run(
        ["node", str(SCRIPT_PATH)],
        input=json.dumps(payload),
        text=True,
        capture_output=True,
        env=env,
        check=False,
    )
    if proc.returncode != 0:
        # Caller owns unlinking the tempfile; surface the helper error.
        detail = (proc.stderr or proc.stdout or "").strip()
        raise RuntimeError(f"browser inspection failed: {detail}")
    report = json.loads(proc.stdout)
    # MJS writes `screenshot_path` (the path it received). Surface
    # it back to the caller so the upload + unlink flow is symmetric.
    report["screenshot_path"] = screenshot_path
    return report


def summary_view(
    report: dict[str, Any],
    *,
    inspection_id: str,
    report_url: str,
    screenshot_url: str,
    scope: str,
    scope_ref: str,
    body_text_limit: int = 4000,
    max_elements: int = 80,
    max_console_messages: int = 50,
    max_network_events: int = 50,
) -> dict[str, Any]:
    """Render the compact summary view of an inspection that becomes the
    MCP tool's return value.

    The summary is intentionally small and bounded: the full structured
    record lives in `report.json` under the artifact URL. This view
    carries the references plus a few key fields a model can act on
    without fetching the artifact.
    """
    body_text_preview = _truncate(str(report.get("body_text") or ""), body_text_limit)
    elements = report.get("elements") or []
    if not isinstance(elements, list):
        elements = []
    elements_preview = list(elements[:max_elements])
    console_messages = report.get("console") or []
    page_errors = report.get("page_errors") or []
    failed_requests = report.get("failed_requests") or []
    http_errors = report.get("http_errors") or []

    def _count(items: Any) -> int:
        return len(items) if isinstance(items, list) else 0

    console_error_count = 0
    if isinstance(console_messages, list):
        console_error_count = sum(1 for m in console_messages if isinstance(m, dict) and m.get("type") == "error")

    return {
        "inspection_id": inspection_id,
        "report_url": report_url,
        "screenshot_url": screenshot_url,
        "scope": scope,
        "scope_ref": scope_ref,
        "final_url": report.get("final_url") or "",
        "status": report.get("status"),
        "title": report.get("title") or "",
        "body_text_preview": body_text_preview,
        "elements_preview": elements_preview,
        "console_messages_preview": _truncate_list(console_messages, max_console_messages),
        "console_error_count": console_error_count,
        "page_error_count": _count(page_errors),
        "page_errors_preview": _truncate_list(page_errors, max_console_messages),
        "http_error_count": _count(http_errors),
        "http_errors_preview": _truncate_list(http_errors, max_network_events),
        "failed_request_count": _count(failed_requests),
        "failed_requests_preview": _truncate_list(failed_requests, max_network_events),
        "inspected_at": report.get("inspected_at") or "",
    }


def _truncate_list(value: Any, limit: int) -> list[Any]:
    if not isinstance(value, list):
        return []
    if limit <= 0:
        return []
    return list(value[:limit])


def fresh_inspection_request_id() -> str:
    """Mint the X-Inspection-Request-Id header value used for idempotent
    retries against POST /v1/inspections."""
    return uuid.uuid4().hex


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Inspect a URL with the leased slot's Playwright and emit JSON.",
    )
    parser.add_argument("url")
    parser.add_argument(
        "--playwright-ws-endpoint",
        required=True,
        help="ws:// URL of the slot's slot-playwright Service",
    )
    parser.add_argument("--width", type=int, default=1440)
    parser.add_argument("--height", type=int, default=900)
    parser.add_argument("--wait-ms", type=int, default=2000)
    parser.add_argument("--timeout-ms", type=int, default=30000)
    parser.add_argument("--no-full-page", action="store_true")
    parser.add_argument("--accessibility", action="store_true")
    parser.add_argument("--max-elements", type=int, default=80)
    parser.add_argument("--body-text-limit", type=int, default=4000)
    args = parser.parse_args()

    result = inspect_url(
        url=args.url,
        playwright_ws_endpoint=args.playwright_ws_endpoint,
        viewport={"width": args.width, "height": args.height},
        wait_ms=args.wait_ms,
        timeout_ms=args.timeout_ms,
        full_page=not args.no_full_page,
        capture_accessibility=args.accessibility,
        max_elements=args.max_elements,
        body_text_limit=args.body_text_limit,
    )
    json.dump(result, sys.stdout, indent=2)
    sys.stdout.write("\n")
    # CLI mode leaves the tempfile on disk — operators invoking the CLI
    # directly typically want to keep the screenshot. Path is in the JSON.


if __name__ == "__main__":
    main()

File: src/mcp_glimmung/browser_inspector.mjs
import { createRequire } from "node:module";
import { pathToFileURL } from "node:url";

const require = createRequire(import.meta.url);
const playwrightModule = pathToFileURL(
  process.env.PLAYWRIGHT_PACKAGE_PATH || require.resolve("playwright"),
).href;
const playwright = await import(playwrightModule);
const chromium = playwright.chromium || playwright.default?.chromium;
if (!chromium) {
  throw new Error(`Unable to load Playwright Chromium from ${playwrightModule}`);
}

const input = JSON.parse(await new Promise((resolve, reject) => {
  let data = "";
  process.stdin.setEncoding("utf8");
  process.stdin.on("data", (chunk) => { data += chunk; });
  process.stdin.on("end", () => resolve(data || "{}"));
  process.stdin.on("error", reject);
}));

if (!input.playwrightWsEndpoint || typeof input.playwrightWsEndpoint !== "string") {
  throw new Error(
    "playwrightWsEndpoint is required: inspect_browser_url runs in the leased "
    + "test slot's slot-playwright pod, not on the MCP host.",
  );
}

function truncate(value, limit) {
  const text = String(value ?? "");
  if (text.length <= limit) return text;
  return `${text.slice(0, Math.max(limit - 3, 0)).trimEnd()}...`;
}

function interestingElements() {
  return (maxElements) => {
    const max = maxElements ?? 80;
    const selectorParts = (el) => {
      if (el.id) return `#${CSS.escape(el.id)}`;
      const testId = el.getAttribute("data-testid") || el.getAttribute("data-test");
      if (testId) return `[data-testid="${CSS.escape(testId)}"]`;
      const aria = el.getAttribute("aria-label");
      if (aria) return `${el.tagName.toLowerCase()}[aria-label="${CSS.escape(aria)}"]`;
      return el.tagName.toLowerCase();
    };
    const selectorFor = (el) => {
      const parts = [];
      let cur = el;
      while (cur && cur.nodeType === Node.ELEMENT_NODE && parts.length < 4) {
        let part = selectorParts(cur);
        if (!cur.id && cur.parentElement) {
          const siblings = Array.from(cur.parentElement.children)
            .filter((s) => s.tagName === cur.tagName);
          if (siblings.length > 1) part += `:nth-of-type(${siblings.indexOf(cur) + 1})`;
        }
        parts.unshift(part);
        if (cur.id) break;
        cur = cur.parentElement;
      }
      return parts.join(" > ");
    };
    const roleFor = (el) => {
      const explicit = el.getAttribute("role");
      if (explicit) return explicit;
      const tag = el.tagName.toLowerCase();
      if (tag === "button") return "button";
      if (tag === "a" && el.hasAttribute("href")) return "link";
      if (tag === "input") return el.getAttribute("type") || "textbox";
      if (/^h[1-6]$/.test(tag)) return "heading";
      if (tag === "select") return "combobox";
      if (tag === "textarea") return "textbox";
      if (tag === "canvas") return "canvas";
      return null;
    };
    const candidates = Array.from(document.querySelectorAll(
      "main,header,nav,section,article,h1,h2,h3,button,a,input,select,textarea,[role],canvas,[data-testid],[data-test]",
    ));
    return candidates
      .map((el) => {
        const rect = el.getBoundingClientRect();
        const style = getComputedStyle(el);
        const visible = rect.width > 0 && rect.height > 0
          && style.visibility !== "hidden"
          && style.display !== "none"
          && Number(style.opacity || "1") > 0;
        const text = (el.innerText || el.getAttribute("aria-label") || el.getAttribute("alt") || "")
          .replace(/\s+/g, " ")
          .trim();
        return {
          selector: selectorFor(el),
          tag: el.tagName.toLowerCase(),
          role: roleFor(el),
          text,
          visible,
          bounds: {
            x: Math.round(rect.x),
            y: Math.round(rect.y),
            width: Math.round(rect.width),
            height: Math.round(rect.height),
          },
          styles: {
            display: style.display,
            position: style.position,
            color: style.color,
            backgroundColor: style.backgroundColor,
            fontSize: style.fontSize,
            zIndex: style.zIndex,
          },
        };
      })
      .filter((el) => el.visible || el.text || el.role || el.tag === "canvas")
      .slice(0, max);
  };
}

function canvasStats() {
  return () => Array.from(document.querySelectorAll("canvas")).map((canvas, index) => {
    const rect = canvas.getBoundingClientRect();
    const out = {
      index,
      selector: canvas.id ? `#${CSS.escape(canvas.id)}` : `canvas:nth-of-type(${index + 1})`,
      width: canvas.width,
      height: canvas.height,
      bounds: {
        x: Math.round(rect.x),
        y: Math.round(rect.y),
        width: Math.round(rect.width),
        height: Math.round(rect.height),
      },
      readable: false,
      nonblank_pixels: null,
      sampled_pixels: null,
    };
    try {
      const ctx = canvas.getContext("2d", { willReadFrequently: true });
      if (!ctx || canvas.width === 0 || canvas.height === 0) return out;
      const sampleW = Math.min(canvas.width, 128);
      const sampleH = Math.min(canvas.height, 128);
      const image = ctx.getImageData(0, 0, sampleW, sampleH).data;
      let nonblank = 0;
      for (let i = 0; i < image.length; i += 4) {
        if (image[i] !== 0 || image[i + 1] !== 0 || image[i + 2] !== 0 || image[i + 3] !== 0) {
          nonblank += 1;
        }
      }
      out.readable = true;
      out.nonblank_pixels = nonblank;
      out.sampled_pixels = sampleW * sampleH;
    } catch (e) {
      out.error = String(e?.message ?? e);
    }
    return out;
  });
}

const startedAt = new Date();
const consoleMessages = [];
const pageErrors = [];
const failedRequests = [];
const httpErrors = [];

const browser = await chromium.connect({ wsEndpoint: input.playwrightWsEndpoint });
try {
  const context = await browser.newContext({
    viewport: input.viewport || { width: 1440, height: 900 },
    ignoreHTTPSErrors: true,
  });

  // Auth-injection plumbing: cookies, extra HTTP headers, and per-origin
  // localStorage seeds. The slot-playwright pod has no credentials of its
  // own — every authenticated browse has to come from the caller, which
  // forwards a session cookie / bearer token / pre-populated storage. See
  // the inspect_url docstring in browser_inspector.py for the typical
  // tank-operator auth.romaine.life → /api/auth/exchange flow that
  // produces these values.
  if (Array.isArray(input.cookies) && input.cookies.length > 0) {
    await context.addCookies(input.cookies);
  }
  if (input.extraHttpHeaders && typeof input.extraHttpHeaders === "object") {
    await context.setExtraHTTPHeaders(input.extraHttpHeaders);
  }
  if (input.localStorage && typeof input.localStorage === "object") {
    // Init scripts run on every page before any page script. We seed
    // localStorage for any origin whose entry matches the current
    // document. Errors (e.g. opaque-origin sandboxed iframes that
    // refuse localStorage access) are swallowed: the goal is best-
    // effort seeding for the top-level navigation, not a guarantee for
    // every nested context.
    await context.addInitScript((items) => {
      try {
        const here = items && items[window.location.origin];
        if (!here) return;
        for (const [k, v] of Object.entries(here)) {
          try { window.localStorage.setItem(k, v); } catch (_) { /* ignore */ }
        }
      } catch (_) { /* ignore */ }
    }, input.localStorage);
  }

  const page = await context.newPage();
  if (input.captureConsole !== false) {
    page.on("console", (msg) => consoleMessages.push({
      type: msg.type(),
      text: truncate(msg.text(), 1000),
      location: msg.location(),
    }));
    page.on("pageerror", (err) => pageErrors.push(String(err?.message ?? err)));
  }
  if (input.captureNetwork !== false) {
    page.on("requestfailed", (req) => failedRequests.push({
      url: req.url(),
      method: req.method(),
      failure: req.failure(),
    }));
    page.on("response", (resp) => {
      if (resp.status() >= 400) {
        httpErrors.push({
          url: resp.url(),
          status: resp.status(),
          status_text: resp.statusText(),
        });
      }
    });
  }

  let status = null;
  let responseError = null;
  let networkIdleReached = false;
  try {
    const response = await page.goto(input.url, {
      waitUntil: "domcontentloaded",
      timeout: input.timeoutMs ?? 30000,
    });
    status = response ? response.status() : null;
  } catch (e) {
    // Navigation itself failed (DNS, refused connection, navigation timeout,
    // etc.). Real error — surface it through response_error.
    responseError = String(e?.message ?? e);
  }
  if (responseError === null) {
    // Navigation succeeded. Wait a bounded window for the page to settle to
    // network idle. Canvas-driven, WebSocket-heavy, polling, and animated
    // pages never reach network idle by design, so timing out here is the
    // expected case rather than an error. Surface it as a soft signal
    // (`network_idle_reached`) and keep response_error clean — callers
    // should not see a "page is broken" signal for a page that's working
    // exactly as intended.
    try {
      await page.waitForLoadState("networkidle", { timeout: Math.min(input.timeoutMs ?? 30000, 10000) });
      networkIdleReached = true;
    } catch (_) {
      networkIdleReached = false;
    }
  }
  if ((input.waitMs ?? 0) > 0) {
    await page.waitForTimeout(input.waitMs);
  }

  let bodyText = "";
  try {
    bodyText = await page.locator("body").innerText({ timeout: 2000 });
  } catch {
    bodyText = "";
  }

  let screenshotPath = null;
  let screenshotBytes = 0;
  if (typeof input.screenshotPath !== "string" || input.screenshotPath.length === 0) {
    throw new Error(
      "screenshotPath is required: the Python wrapper owns the tempfile lifecycle "
      + "and the screenshot is uploaded to glimmung as a durable artifact instead "
      + "of round-tripped through stdout as base64.",
    );
  }
  const fs = await import("node:fs/promises");
  const buf = await page.screenshot({
    path: input.screenshotPath,
    fullPage: input.fullPage !== false,
  });
  screenshotPath = input.screenshotPath;
  screenshotBytes = buf.length;
  // Belt-and-braces: confirm the tempfile got the bytes Chromium just emitted.
  // Playwright writes via stream when `path` is given, but the path side and
  // the buffer return should agree in size.
  const stat = await fs.stat(screenshotPath);
  if (stat.size !== screenshotBytes) {
    throw new Error(
      `screenshot tempfile size=${stat.size} disagrees with chromium buffer size=${screenshotBytes}`,
    );
  }

  let accessibility = null;
  if (input.captureAccessibility) {
    try {
      accessibility = await page.accessibility.snapshot({ interestingOnly: true });
    } catch (e) {
      accessibility = { error: String(e?.message ?? e) };
    }
  }

  const result = {
    schema_version: 2,
    url: input.url,
    final_url: page.url(),
    status,
    response_error: responseError,
    network_idle_reached: networkIdleReached,
    title: await page.title(),
    body_text: truncate(bodyText.replace(/\0/g, ""), input.bodyTextLimit ?? 4000),
    viewport: input.viewport || { width: 1440, height: 900 },
    elements: await page.evaluate(interestingElements(), input.maxElements ?? 80),
    console: consoleMessages,
    page_errors: pageErrors,
    failed_requests: failedRequests,
    http_errors: httpErrors,
    accessibility,
    screenshot_path: screenshotPath,
    screenshot_size_bytes: screenshotBytes,
    canvas: await page.evaluate(canvasStats()),
    inspected_at: startedAt.toISOString(),
  };
  await context.close();
  process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
} finally {
  await browser.close();
}
